M2/Macaulay2/packages/Varieties/SheafMaps.m2

```diff
@@ -0,0 +1,890 @@
+export {
+    -- Types
+    "SheafMap",
+    -- Methods
+    "sheafMap",
+    "isLiftable",
+--    "yonedaSheafExtension",
+    }
+
+-----------------------------------------------------------------------------
+-- SheafHom type declarations and basic constructors
+-----------------------------------------------------------------------------
+
+SheafMap = new Type of HashTable
+SheafMap.synonym = "Morphism of Sheaves"
+
+-- TODO: if over affine variety, dehomogenize the maps
+map(CoherentSheaf, CoherentSheaf, Matrix) := SheafMap => opts -> (G, F, phi) -> (
+    if variety G =!= variety F then error "expected sheaves over the same variety";
+    if not instance(variety F, ProjectiveVariety) then error "maps of sheaves not yet implemented on other varieties";
+    deg := if opts.Degree =!= null then opts.Degree else min flatten degrees source phi;
+    new SheafMap from {
+        symbol source => F,
+        symbol target => G,
+        symbol degree => deg,
+        symbol map => phi,
+        symbol cache => new CacheTable
+        }
+    )
+
+-- when phi is constructed by truncation >= d
+map(CoherentSheaf, CoherentSheaf, Matrix, ZZ) := SheafMap => opts -> (G, F, phi, d) -> (
+    newPhi := inducedMap(module G, target phi) * phi;
+    map(G, F, newPhi, Degree => d))
+map(CoherentSheaf, CoherentSheaf, Matrix, InfiniteNumber) := SheafMap => opts -> (G, F, phi, d) -> (
+    if d === -infinity then map(G, F, phi) else error "unexpected degree for map of sheaves")
+-- TODO: support map(F, F, 1) and map(F, G, 0) for identity and zero maps
+
+sheafMap = method()
+sheafMap Matrix      := SheafMap =>  phi     -> map(sheaf target phi, sheaf source phi, phi)
+sheafMap(Matrix, ZZ) := SheafMap => (phi, d) -> map(sheaf target phi, sheaf source phi, truncate(d, phi), d)
+
+isWellDefined SheafMap := f -> (
+    (G, F) := (target f, source f);
+    X := variety f;
+    d := degree f;
+    all({ G, F, matrix f }, isWellDefined)
+    -- data type checks
+    and condition(set keys f === set { symbol source, symbol target, symbol degree, symbol map, symbol cache },
+	"the hash table does not have the expected keys")
+    and condition(
+	instance(f.source, CoherentSheaf) and
+	instance(f.target, CoherentSheaf) and
+	instance(f.cache, CacheTable) and
+	instance(f.map, Matrix) and
+	instance(f.degree, ZZ),
+	"the hash table does not have the expected values")
+    -- mathematical checks
+    and condition(ring f === ring X,
+	"underlying matrix and variety do not have the same ring")
+    and condition(same {X, variety F, variety G},
+	"underlying variety does not match that of the source and target")
+    and condition(not isProjective X or isHomogeneous matrix f,
+	"underlying matrix of a map of coherent sheaf on a projective variety should be homogeneous")
```

Comment:
Again, is checking ```isHomogeneous``` sufficient.  Which gradings are allowed for projective varieties?